src/led.c

```diff
@@ -166,27 +168,51 @@ void bl_led_status_set(bool enable)
 }
 
 /* Exported function, documented in led.h */
-enum bl_error bl_led_loop(void)
+enum bl_error bl_led_setup(uint16_t led_mask)
 {
-	while (1) {
-		uint16_t led_on = bl_led_active << 1;
-		if (!led_on)
-			led_on = 1;
-
-		if (!(bl_led_mask & led_on))
-		{
-			bl_led_active = led_on;
-			continue;
-		}
+	if (!led_mask) {
+		return BL_ERROR_BAD_SOURCE_MASK;
+	}
+
+	bl_led_active = 0;
+	led_count = 0;
+
+	while (led_mask)
+	{
+		unsigned i = __builtin_ffs(led_mask) - 1;
+		bl_led_src_t *active = &bl_led_src[bl_led_active];
 
-		bl_led__set(LED_PORT_A, led_on);
-		bl_led__set(LED_PORT_B, led_on);
-		bl_led__set(LED_PORT_C, led_on);
+		active->led = i;
+		active->src_mask = (1U << led_table[i].pd | NON_LED_MASKS);
+		active->gpios = (1U << led_table[i].pin);
+		active->gpior = (active->gpios << 16);
+		active->gpio_bsrr = led_port[led_table[i].port_idx] + 0x18;
```

Comment:
Please make a new macro or use an existing one or make a new one. In this case I suspect it'll be something like GPIO_BSRR(led_port[led_table[i].port_idx])

Reply:
GPIO_BSRR de-ref the pointer where we only want to cache the pointer address here. I'll make a new macro for it.